packages/flutter/lib/src/material/tabs.dart

```diff
@@ -945,9 +949,9 @@ class _TabBarState extends State<TabBar> {
     return _tabScrollOffset(_currentIndex, viewportWidth, minExtent, maxExtent);
   }
 
-  void _scrollToCurrentIndex() {
+  void _scrollToCurrentIndex(Duration duration, Curve curve) {
     final double offset = _tabCenteredScrollOffset(_currentIndex);
-    _scrollController.animateTo(offset, duration: kTabScrollDuration, curve: Curves.ease);
+    _scrollController.animateTo(offset, duration: duration ?? kTabScrollDuration, curve: curve ?? Curves.ease);
```

Comment:
_scrollController.animateTo returns a future, we can add a promise after it finishes to set state so that we don't need to use `_handleTabControllerTick` for this.

Reply:
Hm I've tried that, but it seems we have to call setState after `tabController.indexIsChanging` has become false. We call `_scrollController.animateTo` slightly earlier than tabController's `_animationController.animateTo`, so when I try to call setState after above future returns, it emits below error:
![image](https://user-images.githubusercontent.com/4879766/89852195-49208980-dbc9-11ea-8812-6c48eb5abf56.png)

It looks like it's related with using `_ChangeAnimation` and `_DragAnimation` in line 1074 of tabs.dart

Reply:
We need to rebuild right before the animation start to update _DragAnimation into _ChangeAnimation.
Adding the setState prior should does the trick
```
  void _handleTabControllerIndexChangeStarted(int index, Duration duration, Curve curve) {
    if (index != _currentIndex) {
      setState(() {
        _currentIndex = index;
      });
      if (widget.isScrollable)
        _scrollToCurrentIndex(duration, curve);
    }
  }
```

This works because animation always starts on the next frame so the setState will make it just in time to update the _ChangeAnimation at the end of this frame.

The tab highlight on the other hand is a bit weird, it looks like there is some weird issue around how we turn on and off the selected using _buildStyledTab around line 1084 to line 1093. I have not look into too deep what the problem is. but I bet there is something wrong inside how we lerp the text style in line 183

Reply:
I applied the change you suggested, but the above tab text highlighting issue still persists :(

Reply:
Yeah, I think the highlight issue is probably related to how we do the animation lerping in _TabStyle. Sorry i have been a little persist on getting things sorted out. We have put in too much bandage fixes to a point that no one really know how it works exactly :(

It will be great if we can clean it up a little bit now.

Reply:
take the TabController._changeIndex for example. It makes zero sense that we need to notify the listener twice before and after animation ends in line 223 and line 229. The comment does not make sense either. However, that is required to update the _DragAnimation into _ChangeAnimation in TabBar and updates it back from _ChangeAnimation into _DragAnimation.

Reply:
No problem :) I understand sorting things out is really important. I also agree it needs cleanup, but maybe that should be done apart from this PR? What do you think?

Reply:
yes, we don't need to clean up everything in this pr, i think the last thing we need to figure out to merge this pr is what cause the highlight issue, and if we can fix it without _handleTabControllerTick. I am ok with using _handleTabControllerTick if it is unavoidable as long as we know the reason behind it.